{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from code_kernel.ipynb\n",
      "importing Jupyter notebook from code_kwargs.ipynb\n",
      "importing Jupyter notebook from code_utils.ipynb\n",
      "importing Jupyter notebook from code_heatmaps.ipynb\n",
      "importing Jupyter notebook from code_bar.ipynb\n",
      "importing Jupyter notebook from code_scatter.ipynb\n",
      "importing Jupyter notebook from code_PCA.ipynb\n",
      "importing Jupyter notebook from code_other.ipynb\n",
      "importing Jupyter notebook from code_plotly.ipynb\n",
      "importing Jupyter notebook from code_pymol.ipynb\n"
     ]
    }
   ],
   "source": [
    "# Regular libraries\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "# Local imports\n",
    "try: # Work with .py files\n",
    "    from mutagenesis_visualization.main.scripts.code_kernel import (\n",
    "        plot_kernel,\n",
    "        plot_hist,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_heatmaps import (\n",
    "        plot_heatmap,\n",
    "        plot_heatmap_rows,\n",
    "        plot_heatmap_columns,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_bar import (\n",
    "        plot_mean,\n",
    "        plot_meandifferential,\n",
    "        plot_position,\n",
    "        plot_meancounts,\n",
    "        plot_library_representation,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_scatter import (\n",
    "        plot_scatter,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_PCA import (\n",
    "        plot_correlation,\n",
    "        plot_individual_correlation,\n",
    "        plot_group_correlation,\n",
    "        plot_pca,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_other import (\n",
    "        plot_rank,\n",
    "        plot_miniheatmap,\n",
    "        plot_neighboreffect,\n",
    "        plot_secondary,\n",
    "        plot_roc,\n",
    "        plot_cumulative,\n",
    "    )\n",
    "\n",
    "    from mutagenesis_visualization.main.scripts.code_plotly import (\n",
    "        plot_mean_plotly,\n",
    "        plot_heatmap_plotly,\n",
    "        plot_histogram_plotly,\n",
    "        plot_rank_plotly,\n",
    "        plot_scatter_plotly,\n",
    "        plot_scatter_3D_plotly,\n",
    "        plot_scatter_3D_pdbprop_plotly,\n",
    "    )\n",
    "    from mutagenesis_visualization.main.scripts.code_utils import (\n",
    "        _is_DNA, _translate_codons, _transform_sequence, _transform_dataset,\n",
    "        _select_SNV, _select_nonSNV, _transform_secondary\n",
    "    )\n",
    "except ModuleNotFoundError: # Work on jupyter notebook\n",
    "    import import_notebook\n",
    "    from code_kernel import (plot_kernel, plot_hist)\n",
    "    from code_heatmaps import (\n",
    "        plot_heatmap,\n",
    "        plot_heatmap_rows,\n",
    "        plot_heatmap_columns,\n",
    "    )\n",
    "    from code_bar import (\n",
    "        plot_mean,\n",
    "        plot_meandifferential,\n",
    "        plot_position,\n",
    "        plot_meancounts,\n",
    "        plot_library_representation,\n",
    "    )\n",
    "    from code_scatter import (plot_scatter)\n",
    "    from code_PCA import (\n",
    "        plot_correlation,\n",
    "        plot_individual_correlation,\n",
    "        plot_group_correlation,\n",
    "        plot_pca,\n",
    "    )\n",
    "    from code_other import (\n",
    "        plot_rank,\n",
    "        plot_miniheatmap,\n",
    "        plot_neighboreffect,\n",
    "        plot_secondary,\n",
    "        plot_roc,\n",
    "        plot_cumulative,\n",
    "    )\n",
    "\n",
    "    from code_plotly import (\n",
    "        plot_mean_plotly,\n",
    "        plot_heatmap_plotly,\n",
    "        plot_histogram_plotly,\n",
    "        plot_rank_plotly,\n",
    "        plot_scatter_plotly,\n",
    "        plot_scatter_3D_plotly,\n",
    "        plot_scatter_3D_pdbprop_plotly,\n",
    "    )\n",
    "    from code_utils import (\n",
    "        _is_DNA,\n",
    "        _translate_codons,\n",
    "        _transform_sequence,\n",
    "        _transform_dataset,\n",
    "        _select_SNV,\n",
    "        _select_nonSNV,\n",
    "        _transform_secondary,\n",
    "    )\n",
    "\n",
    "try:\n",
    "    from mutagenesis_visualization.main.scripts.code_pymol import (plot_pymol)\n",
    "except ModuleNotFoundError:\n",
    "    from code_pymol import (plot_pymol)\n",
    "except ModuleNotFoundError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Counts:\n",
    "    """\n",
    "    *Counts* represents the output of reading a fastq file.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pandas dataframe\n",
    "        Dataframe containing the counts per codon.\n",
    "\n",
    "    start_position : int, default None\n",
    "        First position in the protein sequence that will be used for the first column of the\n",
    "        array. If a protein has been mutated only from residue 100-150, then if start_position = 100,\n",
    "        the algorithm will trim the first 99 amino acids in the input sequence. The last \n",
    "        residue will be calculated based on the length of the input array. We have set the default value to 2\n",
    "        because normally the Methionine in position 1 is not mutated.\n",
    "\n",
    "    aminoacids : list, default None\n",
    "        List of aminoacids (in order). Stop codon needs to be '*'.\n",
    "        If none, it will use the index of the dataframe   \n",
    "\n",
    "    """\n",
    "    def __init__(self, df, start_position=None, aminoacids=None):\n",
    "        self.dataframe = df\n",
    "\n",
    "        if start_position:\n",
    "            self.start_position = start_position\n",
    "            self.positions = np.arange(start_position, len(df.columns) + 1)\n",
    "        else:  # if none, use the columns of the dataframe\n",
    "            self.positions = list(df.columns)\n",
    "            self.start_position = self.positions[0]\n",
    "\n",
    "        if aminoacids:\n",
    "            self.aminoacids = aminoacids\n",
    "        else:  # if aminoacids is none, use the index of the dataframe\n",
    "            if _is_DNA(df):\n",
    "                self.aminoacids = _translate_codons(df)\n",
    "            else:\n",
    "                self.aminoacids = list(df.index)\n",
    "        # bar plots\n",
    "\n",
    "    mean_counts = plot_meancounts\n",
    "    library_representation = plot_library_representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.643452Z",
     "start_time": "2020-10-06T23:17:05.633020Z"
    }
   },
   "outputs": [],
   "source": [
    "class Screen:\n",
    "    """\n",
    "    *Screen* represents a mutagenesis experiment. If you are doing deep scan \n",
    "    mutagenesis, then every amino acid in the protein has been mutated to every possible \n",
    "    amino acid. For example, if there was a leucine at position 2, then this leucine would\n",
    "    be mutated to the other 19 naturally occurring amino acids. However, you can also use the \n",
    "    package if you only have a handful of amino acid substitutions.\n",
    "\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dataset : array\n",
    "        2D matrix containing the enrichment scores of the point mutants. Columns will contain the\n",
    "        amino acid substitutions, rows will contain the enrichment for each residue in the protein sequence.\n",
    "\n",
    "    sequence : str\n",
    "        Protein sequence (columns) in 1 letter code format.\n",
    "\n",
    "    aminoacids : list, default list('ACDEFGHIKLMNPQRSTVWY*')\n",
    "        Amino acid substitutions (rows). Submit in the same order that is used for the array.\n",
    "\n",
    "    start_position : int, default 2\n",
    "        First position in the protein sequence that will be used for the first column of the\n",
    "        array. If a protein has been mutated only from residue 100-150, then if start_position = 100,\n",
    "        the algorithm will trim the first 99 amino acids in the input sequence. The last \n",
    "        residue will be calculated based on the length of the input array. We have set the default value to 2\n",
    "        because normally the Methionine in position 1 is not mutated.\n",
    "\n",
    "    secondary : list, optional\n",
    "        This parameter is used to group the data by secondary structure. The format is \n",
    "        the name of the secondary structure multiplied by the residue length of that motif.\n",
    "        Example : [['β1']*(8),['L1']*(7),['α1']*(9),...,].\n",
    "\n",
    "    roc_df: Pandas dataframe, optional\n",
    "        A dataframe that contains a column of variants labeled 'Variant' with a column labeled 'Class'\n",
    "        containing the true class of that mutation. This can be used to compare enrichment scores to some label (such as \n",
    "        pathogenicity as found in a Cancer database) using ROC AUC.\n",
    "\n",
    "    fillna : int, default 0\n",
    "        How to replace NaN values.\n",
    "\n",
    "\n",
    "    Attributes\n",
    "    ------------\n",
    "    dataframe : pandas dataframe\n",
    "        Contains the enrichment scores, position, sequence.\n",
    "\n",
    "    Other attributes are same as input parameters: dataset, aminoacids, start_position, roc_df, secondary\n",
    "\n",
    "    """\n",
    "    def __init__(\n",
    "        self,\n",
    "        dataset,\n",
    "        sequence,\n",
    "        aminoacids=list('ACDEFGHIKLMNPQRSTVWY*'),\n",
    "        start_position=2,\n",
    "        fillna=0,\n",
    "        secondary=None,\n",
    "        roc_df=None\n",
    "    ):\n",
    "        # Instances\n",
    "        self.dataset = np.array(dataset)\n",
    "        self.aminoacids = aminoacids\n",
    "        self.start_position = start_position\n",
    "        self.end_position = len(self.dataset[0]) + start_position\n",
    "        self.sequence_raw = ''.join(sequence)  # why I am doing this?\n",
    "        self.sequence = _transform_sequence(\n",
    "            self.dataset, self.sequence_raw, self.start_position\n",
    "        )\n",
    "        self.dataframe_stopcodons, self.dataframe = _transform_dataset(\n",
    "            self.dataset, self.sequence, self.aminoacids, self.start_position,\n",
    "            fillna\n",
    "        )\n",
    "        self.dataframe_SNV = _select_SNV(self.dataframe)\n",
    "        self.dataframe_nonSNV = _select_nonSNV(self.dataframe)\n",
    "\n",
    "        # Optional parameters\n",
    "        self.roc_df = roc_df\n",
    "        self.secondary = secondary\n",
    "        if self.secondary is not None:\n",
    "            self.secondary, self.secondary_dup = _transform_secondary(\n",
    "                self.dataset, self.secondary, self.start_position,\n",
    "                self.aminoacids\n",
    "            )\n",
    "\n",
    "        # Assert messages\n",
    "        assert len(sequence) >= len(\n",
    "            self.dataset[0]\n",
    "        ), 'Input sequence is not long enough'\n",
    "\n",
    "    # Methods (Associated functions)\n",
    "\n",
    "    # code_kernel\n",
    "    kernel = plot_kernel\n",
    "    histogram = plot_hist\n",
    "\n",
    "    # code_heatmaps\n",
    "    heatmap = plot_heatmap\n",
    "    heatmap_rows = plot_heatmap_rows\n",
    "    heatmap_columns = plot_heatmap_columns\n",
    "\n",
    "    # code_bar\n",
    "    mean = plot_mean\n",
    "    differential = plot_meandifferential\n",
    "    position = plot_position\n",
    "    meancounts = plot_meancounts\n",
    "    library_representation = plot_library_representation\n",
    "\n",
    "    # code_scatter\n",
    "    scatter = plot_scatter\n",
    "\n",
    "    # code_PCA\n",
    "    correlation = plot_correlation\n",
    "    individual_correlation = plot_individual_correlation\n",
    "    group_correlation = plot_group_correlation\n",
    "    pca = plot_pca\n",
    "\n",
    "    # code_Other\n",
    "    rank = plot_rank\n",
    "    miniheatmap = plot_miniheatmap\n",
    "    neighboreffect = plot_neighboreffect\n",
    "    secondary_mean = plot_secondary\n",
    "    roc = plot_roc\n",
    "    cumulative = plot_cumulative\n",
    "\n",
    "    # code_plotly\n",
    "    rank_plotly = plot_rank_plotly\n",
    "    scatter_plotly = plot_scatter_plotly\n",
    "    histogram_plotly = plot_histogram_plotly\n",
    "    mean_plotly = plot_mean_plotly\n",
    "    scatter_3D_plotly = plot_scatter_3D_plotly\n",
    "    scatter_3D_pdbprop_plotly = plot_scatter_3D_pdbprop_plotly\n",
    "    heatmap_plotly = plot_heatmap_plotly\n",
    "\n",
    " \n",
    "    # pymol\n",
    "    try:\n",
    "        pymol = plot_pymol\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
