{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    },
    "execution": {
     "iopub.status.idle": "2020-10-11T02:23:58.137436Z",
     "shell.execute_reply": "2020-10-11T02:23:58.136608Z",
     "shell.execute_reply.started": "2020-10-11T02:23:53.211168Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import copy\n",
    "from scipy import stats\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from sklearn import metrics\n",
    "import freesasa\n",
    "from os import path\n",
    "import os\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "\n",
    "try:\n",
    "    import shannon\n",
    "except ModuleNotFoundError:\n",
    "    pass\n",
    "    \n",
    "\n",
    "# kwargs and parameters\n",
    "try:\n",
    "    import import_notebook\n",
    "except ModuleNotFoundError:\n",
    "    pass\n",
    "\n",
    "import code_kwargs\n",
    "import code_utils\n",
    "from code_heatmaps import _labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.124811Z",
     "start_time": "2020-10-06T23:17:05.103434Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.461160Z",
     "iopub.status.busy": "2020-10-11T02:23:58.460923Z",
     "iopub.status.idle": "2020-10-11T02:23:58.471800Z",
     "shell.execute_reply": "2020-10-11T02:23:58.470626Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.461143Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_rank(self, mode='pointmutant', outdf=False, \n",
    "              output_file: Union[None, str, Path] = None,**kwargs):\n",
    "    '''\n",
    "    DEPRECATED\n",
    "    Generate a rank plot so every mutation/residue is sorted based on enrichment score.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "        \n",
    "    mode : str, default 'pointmutant'. \n",
    "        Alternative set to \"mean\" for the mean of each position\n",
    "    \n",
    "    outdf : boolean, default False\n",
    "        If set to true, will return the df with the rank of mutations\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                        \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    Pandas dataframe\n",
    "    '''\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (4, 2))\n",
    "    temp_kwargs['x_label'] = kwargs.get('x_label', 'Rank')\n",
    "    temp_kwargs['y_label'] = kwargs.get('y_label', r'$∆E^i_x$')\n",
    "    \n",
    "    # Sort by enrichment scores\n",
    "    df = self.dataframe.sort_values(by=['Score']).copy()\n",
    "    \n",
    "    # Chose mode:\n",
    "    if mode == 'mean':\n",
    "        df = df.groupby(by=['Position'],as_index=False).mean()\n",
    "        df.sort_values(by=['Score'], inplace=True)\n",
    "\n",
    "    # create figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "\n",
    "    # import parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # Scatter data points\n",
    "    plt.scatter(np.arange(len(df),0,-1), df['Score'], c='k', s=1)\n",
    "\n",
    "    # Titles\n",
    "    plt.title(temp_kwargs['title'], fontsize=12,\n",
    "              fontname='Arial', color='k', pad=8)\n",
    "    # Labels\n",
    "    plt.ylabel(temp_kwargs['y_label'], fontsize=10,\n",
    "               fontname=\"Arial\", color='k', labelpad=0)\n",
    "    plt.xlabel(temp_kwargs['x_label'], fontsize=10,\n",
    "               fontname=\"Arial\", color='k')\n",
    "    \n",
    "    # other graph parameters\n",
    "    plt.xlim(temp_kwargs['xscale'])\n",
    "    plt.ylim(temp_kwargs['yscale'])\n",
    "    \n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    \n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    if outdf:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Miniheatmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mean substitution heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.181693Z",
     "start_time": "2020-10-06T23:17:05.163286Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.504614Z",
     "iopub.status.busy": "2020-10-11T02:23:58.504059Z",
     "iopub.status.idle": "2020-10-11T02:23:58.518852Z",
     "shell.execute_reply": "2020-10-11T02:23:58.517809Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.504586Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_miniheatmap(self, offset=0, output_file: Union[None, str, Path] = None,\n",
    "                     **kwargs):\n",
    "    '''\n",
    "    Generate a miniheatmap plot enrichment scores of mutagenesis selection assays.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    offset : int, default 0\n",
    "        if you want to study effects of a residue when is behind or in front of another residue.\n",
    "        offset of 1 means that you evaluate the effect of following residue n+1 on n.\n",
    "        offset of -1 means that you look at the previous residue (n-1 on n).\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "\n",
    "    # load labels\n",
    "    temp_kwargs['color_sequencelabels'] = _labels(self.start_position)[0]\n",
    "    temp_kwargs['number_sequencelabels'] = _labels(self.start_position)[1]\n",
    "\n",
    "    # do offset if appropriate\n",
    "    dataframe_stopcodons = _transform_dataset_offset(self, offset)\n",
    "\n",
    "    # calculate condensed heatmap\n",
    "    dataset = _condense_heatmap(\n",
    "        dataframe_stopcodons, temp_kwargs['neworder_aminoacids'])\n",
    "\n",
    "    _plot_miniheatmap(dataset, output_file, temp_kwargs)\n",
    "\n",
    "    return\n",
    "\n",
    "\n",
    "def _condense_heatmap(df, new_order):\n",
    "    '''\n",
    "    Converts the np.array with stored enrichment scores into the condensed heatmap\n",
    "    '''\n",
    "    # Convert dataset to df\n",
    "    df = df.copy()\n",
    "    df.drop(['Position'], axis=1, inplace=True)\n",
    "\n",
    "    # Group by sequence and aminoacid, and then pivot table\n",
    "    df_grouped = df.groupby(['Sequence', 'Aminoacid'], sort = False).mean()\n",
    "    df_pivoted = df_grouped.pivot_table(values='Score',\n",
    "                                        index='Aminoacid',  columns='Sequence')\n",
    "    df_pivoted.reset_index(drop=False, inplace=True)\n",
    "\n",
    "    # Sort in y axis desired order\n",
    "    df_pivoted['Aminoacid'] = pd.Categorical(\n",
    "        df_pivoted['Aminoacid'], new_order)\n",
    "    df_pivoted = df_pivoted.sort_values(by=['Aminoacid'])\n",
    "\n",
    "    # Sort in x axis desired order\n",
    "    x_order = code_utils._common(new_order, list(df_pivoted.columns))\n",
    "\n",
    "    # Drop amino acid column\n",
    "    data_dropped = df_pivoted.drop(['Aminoacid'], axis=1)\n",
    "\n",
    "    return data_dropped[x_order]\n",
    "\n",
    "\n",
    "def _offset_sequence(dataset, sequence, start_position, offset):\n",
    "    '''\n",
    "    Internal function that offsets the input sequence\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dataset, sequence, start_position, offset\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    string containing trimmed sequence\n",
    "    '''\n",
    "    # Deep copy sequence\n",
    "    sequence = copy.deepcopy(sequence)\n",
    "\n",
    "    # truncate sequence\n",
    "    if offset > 0:\n",
    "        sequence = sequence+'X'*np.absolute(offset)\n",
    "        trimmedsequence = sequence[start_position-1 +\n",
    "                                   offset:len(dataset[0])+start_position-1+offset]\n",
    "    else:\n",
    "        sequence = 'X'*(np.absolute(offset))+sequence\n",
    "        trimmedsequence = sequence[start_position -\n",
    "                                   1:len(dataset[0])+start_position-1]\n",
    "\n",
    "    return trimmedsequence\n",
    "\n",
    "\n",
    "def _transform_dataset_offset(self, offset, stopcodons=True):\n",
    "    '''\n",
    "    Generate a dataframe with the sequence offset. Reutilizes _transform_dataset\n",
    "    '''\n",
    "    # Add offset sequence\n",
    "    offset_sequence = _offset_sequence(self.dataset, self.sequence_raw,\n",
    "                                       self.start_position, offset)\n",
    "    df = self.dataframe_stopcodons.copy() if stopcodons is True else self.dataframe.copy()\n",
    "\n",
    "    # Copy old sequence\n",
    "    df['Sequence_old'] = df['Sequence']\n",
    "    # Count amino acids\n",
    "    aa_number = len(set(df['Aminoacid']))\n",
    "    # Generate new offset sequence\n",
    "    df['Sequence'] = np.ravel([[aa]*aa_number for aa in offset_sequence])\n",
    "\n",
    "    # Drop rows with X\n",
    "    df.drop(df.index[df['Sequence'] == 'X'], inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neighbor residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.213829Z",
     "start_time": "2020-10-06T23:17:05.186789Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.520766Z",
     "iopub.status.busy": "2020-10-11T02:23:58.520410Z",
     "iopub.status.idle": "2020-10-11T02:23:58.540712Z",
     "shell.execute_reply": "2020-10-11T02:23:58.540144Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.520736Z"
    }
   },
   "outputs": [],
   "source": [
    " def plot_neighboreffect (self, offset=1, output_file: Union[None, str, Path] = None,\n",
    "                          **kwargs):\n",
    "    '''\n",
    "    Generate a miniheatmap plot telling you the effect of having a residue in front or behind.\n",
    "    It corrects for the effect of that amino acid on the rest of the population.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    offset : int, default 1\n",
    "        if you want to study effects of a residue when is behind or in front of another residue.\n",
    "        offset of 1 means that you evaluate the effect of following residue n+1 on n. On a \"MTEY...\" sequence,\n",
    "        you would look at the effect of T on M, E on T, Y on E, etc.. and then group by residue (n+1).\n",
    "        offset of -1 means that you look at the previous residue (n-1 on n).\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    if '*' in temp_kwargs['neworder_aminoacids']: temp_kwargs['neworder_aminoacids'].remove('*')\n",
    "    \n",
    "    # do offset, no stop codons\n",
    "    df = _normalize_neighboreffect(self,offset,temp_kwargs['neworder_aminoacids'])\n",
    "    \n",
    "    # Plot\n",
    "    _plot_miniheatmap(df,temp_kwargs)\n",
    "    \n",
    "    return\n",
    "    \n",
    "def _plot_miniheatmap(df,output_file, temp_kwargs):\n",
    "    # declare figure and subplots\n",
    "    coeff = len(df.columns)/19*1.05\n",
    "    fig = plt.figure(figsize=(2.5*coeff, 2.5))\n",
    "    gs = gridspec.GridSpec(nrows=1, ncols=1)\n",
    "    ax = plt.subplot(gs[0])\n",
    "\n",
    "    # main heatmap\n",
    "    heatmap = ax.pcolor(df.to_numpy(), vmin=temp_kwargs['colorbar_scale'][0], vmax=temp_kwargs['colorbar_scale'][1],\n",
    "                        cmap=temp_kwargs['colormap'], edgecolors='k', linewidths=0.2, color='darkgrey')\n",
    "\n",
    "    # ____________axes manipulation____________________________________________\n",
    "    # put the major ticks at the middle of each cell\n",
    "    ax.set_xticks(np.arange(df.shape[1]) + 0.5, minor=False)\n",
    "    ax.set_yticks(np.arange(df.shape[0]) + 0.5, minor=False)\n",
    "\n",
    "    # position of axis labels\n",
    "    ax.tick_params('x', direction='out', pad=-2.5)\n",
    "    ax.tick_params('y', direction='out', pad=0.4)\n",
    "\n",
    "    # want a more natural, table-like display\n",
    "    ax.invert_yaxis()\n",
    "    ax.xaxis.tick_top()\n",
    "\n",
    "    # remove ticks\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "\n",
    "    # so labels of x and y do not show up and my labels show up instead\n",
    "    ax.set_xticklabels(list(df.columns), fontsize=6.5,\n",
    "                       fontname=\"Arial\", color='k', minor=False)\n",
    "    ax.set_yticklabels(temp_kwargs['neworder_aminoacids'],\n",
    "                       fontsize=6.5, fontname=\"Arial\", color='k', minor=False)\n",
    "\n",
    "    # align the labels of the y axis\n",
    "    for ylabel in ax.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "\n",
    "    # _____________________________________________________________________________\n",
    "\n",
    "    # for color bar format\n",
    "    cb = plt.colorbar(heatmap, fraction=0.025, pad=0.05, aspect=5, ticks=[temp_kwargs['colorbar_scale'][0], np.mean(\n",
    "        temp_kwargs['colorbar_scale']), temp_kwargs['colorbar_scale'][1]], orientation='vertical')\n",
    "    cb.ax.set_yticklabels(cb.ax.get_yticklabels(), fontsize=7, fontname=\"Arial\", color='k')\n",
    "    cb.update_ticks()\n",
    "    plt.text(len(df.columns)+2, 7.8, r'$\\langle∆E^x_i\\rangle_x$', horizontalalignment='center',\n",
    "             fontsize=7, fontname=\"Arial\", color='k')\n",
    "\n",
    "    # for putting title on graph\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=10)\n",
    "    plt.ylabel('Amino Acid Substitution', fontsize=10, labelpad=-1)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    \n",
    "    if temp_kwargs['show']: plt.show()\n",
    "    return\n",
    "\n",
    "def _normalize_neighboreffect(self,offset,neworder):\n",
    "    '''\n",
    "    For every residue, subtract the average effect of a substitution\n",
    "    Returns a normalized dataframe\n",
    "    '''\n",
    "    aalist = list('ACDEFGHIKLMNPQRSTVWY')\n",
    "    # Add offset sequence to df\n",
    "    df = _transform_dataset_offset(self,offset,False)\n",
    "    \n",
    "    # calculate mean effect using condensed heatmap\n",
    "    mean = _condense_heatmap(self.dataframe, aalist)\n",
    "    \n",
    "    df_normalized = pd.DataFrame()\n",
    "    for aa in aalist:\n",
    "        # Choose the neighbors of an aa\n",
    "        aa_neighbors = df.loc[df['Sequence']==aa]\n",
    "        # Do the mean substitution of amino acids that are repeated\n",
    "        aa_neighbors = aa_neighbors.groupby(['Sequence_old','Aminoacid'],as_index=False).mean()\n",
    "        # Make into table\n",
    "        aa_neighbors_pivoted = aa_neighbors.pivot_table(values='Score', index='Aminoacid',  columns='Sequence_old')\n",
    "        aa_neighbors_pivoted.reset_index(drop=True, inplace=True)\n",
    "        # Get the mean of the amino acids that appear in the aa_neighbors subset\n",
    "        mean_neighbors = mean[list(aa_neighbors_pivoted.columns)]\n",
    "        # Subtract average effect and do mean\n",
    "        df_normalized[aa] = (aa_neighbors_pivoted - mean_neighbors).mean(axis=1)\n",
    "    \n",
    "    # Sort by aa\n",
    "    df_normalized = df_normalized[neworder]\n",
    "    # Sort in y axis desired order\n",
    "    df_normalized = _sort_yaxis_aminoacids(df_normalized,neworder,aalist)\n",
    "    return df_normalized\n",
    "\n",
    "def _sort_yaxis_aminoacids(df,neworder,oldorder=list('ACDEFGHIKLMNPQRSTVWY')):\n",
    "    # Sort in y axis desired order\n",
    "    df['Aminoacid_new'] = oldorder\n",
    "    df['Aminoacid_new'] = pd.Categorical(df['Aminoacid_new'], neworder)\n",
    "    df.sort_values(by=['Aminoacid_new'],inplace=True)\n",
    "    df.drop(['Aminoacid_new'], inplace=True, axis=1)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Secondary Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.416985Z",
     "start_time": "2020-10-06T23:17:05.401848Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.618542Z",
     "iopub.status.busy": "2020-10-11T02:23:58.617358Z",
     "iopub.status.idle": "2020-10-11T02:23:58.629589Z",
     "shell.execute_reply": "2020-10-11T02:23:58.629023Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.618255Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_secondary(self, output_file: Union[None, str, Path] = None, **kwargs):\n",
    "    '''\n",
    "    Generates a bar plot of data sorted by secondary elements (alpha helices and beta sheets).\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "\n",
    "    '''\n",
    "    # Load parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # Update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (2.5, 2))\n",
    "    temp_kwargs['yscale'] = kwargs.get('yscale', (-2, 1))\n",
    "\n",
    "    # Get data\n",
    "    df = _calculate_secondary(self.dataframe, self.secondary_dup)\n",
    "\n",
    "    # Color\n",
    "    df['Color'] = df.apply(code_utils._color_data, axis=1,\n",
    "                           args=(temp_kwargs['color_gof'],\n",
    "                                 temp_kwargs['color_lof']))\n",
    "\n",
    "    # Make figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    ticks = np.arange(0, len(df))  # label locations\n",
    "    width = 0.5\n",
    "    labels = df['Secondary']\n",
    "\n",
    "    # Plot figure\n",
    "    ax.bar(ticks, df['Score'], width, color=df['Color'], ec='k',)\n",
    "\n",
    "    # graph parameters\n",
    "    ax.set_xticks(ticks)\n",
    "    ax.set_xticklabels(labels, fontsize=9, fontname=\"Arial\",\n",
    "                       color='k', minor=False, rotation=0)\n",
    "    ax.set_ylabel(r'$∆E^i_x$', fontsize=10, fontname=\"Arial\",\n",
    "                  color='k', labelpad=12, rotation=0)\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=5)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _calculate_secondary(df, secondary):\n",
    "    '''\n",
    "    Returns copy\n",
    "    '''\n",
    "    df = df.copy()\n",
    "    df.insert(4, 'Secondary', secondary)\n",
    "    df = df.groupby(['Secondary'], as_index=False, sort=False).mean()\n",
    "    df = df[df['Secondary'].str.startswith(('β', 'α'))]\n",
    "    df = df.drop(['Position'], axis=1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ROC AUC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.447269Z",
     "start_time": "2020-10-06T23:17:05.420206Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.631812Z",
     "iopub.status.busy": "2020-10-11T02:23:58.631556Z",
     "iopub.status.idle": "2020-10-11T02:23:58.647916Z",
     "shell.execute_reply": "2020-10-11T02:23:58.646729Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.631782Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_roc(self, df_class=None, output_file: Union[None, str, Path] = None,\n",
    "             **kwargs):\n",
    "    '''\n",
    "    Generates ROC AUC plot. It compares enrichment scores to some labels that the user has specified.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    df_class: Pandas dataframe\n",
    "        A dataframe that contains a column of variants labeled 'Variant' with a column labeled 'Class'\n",
    "        containing the true class of that mutation. The true class can also be an input when creating the object.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                \n",
    "    Returns\n",
    "    --------\n",
    "    None.\n",
    "    '''\n",
    "    # Use default class\n",
    "    if df_class is None:\n",
    "        df_class = self.roc_df\n",
    "\n",
    "    # Merge dataframe with classes\n",
    "    df = _mergeclassvariants(df_class, self.dataframe)\n",
    "\n",
    "    # Calculate ROC parameters\n",
    "    fpr, tpr, auc, _ = _rocauc(df)\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (2.5, 2.5))\n",
    "\n",
    "    # import parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # create figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    lw = 2\n",
    "    plt.plot(fpr, tpr, color='k', lw=lw, label='AUC = %0.2f' % auc)\n",
    "    plt.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')\n",
    "\n",
    "    # Graph limits\n",
    "    plt.xlim([0.0, 1.0])\n",
    "    plt.ylim([0.0, 1.05])\n",
    "    tick_spacing = 0.2\n",
    "    ax.xaxis.set_major_locator(\n",
    "        ticker.MultipleLocator(tick_spacing))  # Plt ticks\n",
    "    ax.yaxis.set_major_locator(\n",
    "        ticker.MultipleLocator(tick_spacing))  # Plt ticks\n",
    "\n",
    "    # Axis labels\n",
    "    plt.title(temp_kwargs['title'], fontsize=12, fontname='Arial', color='k')\n",
    "    plt.ylabel('True Positive Rate', fontsize=12,\n",
    "               fontname=\"Arial\", color='k', labelpad=0)\n",
    "    plt.xlabel('False Positive Rate', fontsize=12, fontname=\"Arial\", color='k')\n",
    "\n",
    "    # Legend\n",
    "    plt.legend(loc='lower right', handlelength=0,\n",
    "               handletextpad=0, frameon=False)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _rocauc(df):\n",
    "    '''\n",
    "    Calculate roc rates and auc.\n",
    "\n",
    "    The input is a dataframe that contains [Variants,Class,Score]\n",
    "    '''\n",
    "    fpr, tpr, thresholds = metrics.roc_curve(\n",
    "        df['Class'], df['Score'], drop_intermediate=True)\n",
    "    auc = metrics.roc_auc_score(df['Class'], df['Score'])\n",
    "    return fpr, tpr, auc, thresholds\n",
    "\n",
    "\n",
    "def _mergeclassvariants(df_score, df_class):\n",
    "    '''\n",
    "    Merge the input dataframe containing the class (true score) for variants and the enrichment scores\n",
    "    '''\n",
    "    # Merge DMS with true score dataset\n",
    "    df_merged = pd.merge(df_class, df_score, on=['Variant'], how='left')\n",
    "\n",
    "    # Drop rows with Nan values\n",
    "    df_merged.dropna(inplace=True)\n",
    "\n",
    "    return df_merged\n",
    "\n",
    "\n",
    "def _concattrueposneg(df_tp, df_tn, subset='Variant', keep='first'):\n",
    "    '''\n",
    "    Concat a df containing the true positive variants and the true negative variants\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    df_tp : Dataframe with the true positives\n",
    "    df_tn : Dataframe with the true negatives\n",
    "    subset : str, default Variant\n",
    "    keep : {‘first’, ‘last’, False} \n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    None.\n",
    "    '''\n",
    "    # Concatenate tp and tn datasets\n",
    "    df_true = pd.concat([df_tp, df_tn], sort=False)\n",
    "\n",
    "    # Will keep a variant as true positive if found in both datasets (because could be a mistake in gnomAD)\n",
    "    df_true.drop_duplicates(subset=subset, keep=keep, inplace=True)\n",
    "\n",
    "    return df_true"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cumulative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.474047Z",
     "start_time": "2020-10-06T23:17:05.457577Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.651159Z",
     "iopub.status.busy": "2020-10-11T02:23:58.650193Z",
     "iopub.status.idle": "2020-10-11T02:23:58.664055Z",
     "shell.execute_reply": "2020-10-11T02:23:58.662916Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.651121Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_cumulative(self, mode='all', output_file: Union[None, str, Path] = None,\n",
    "                    **kwargs):\n",
    "    '''\n",
    "    Generates a cumulative plot of the enrichment scores by position. \n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    mode : str, default 'all' \n",
    "        Options are 'all','SNV' and 'nonSNV'.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    None.\n",
    "    '''\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (3, 2))\n",
    "    temp_kwargs['tick_spacing'] = kwargs.get('tick_spacing', 20)\n",
    "\n",
    "    # import parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # create figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "\n",
    "    # Get data filtered\n",
    "    df = _filter_bySNV(self, mode)\n",
    "    cumsum = df.cumsum(skipna=False)['Score']\n",
    "    plt.plot(df['Position'], cumsum/list(cumsum)[-1], color='red', lw=2)\n",
    "\n",
    "    # y label\n",
    "    y_label = 'Cumulative LoF'\n",
    "    if list(cumsum)[-1] > 0:\n",
    "        y_label = 'Cumulative GoF'\n",
    "\n",
    "    # Graph limits\n",
    "    plt.xlim(self.dataframe['Position'].min(),\n",
    "             self.dataframe['Position'].max()+1)\n",
    "    plt.ylim(0, 1.1)\n",
    "\n",
    "    ax.xaxis.set_major_locator(ticker.MultipleLocator(\n",
    "        temp_kwargs['tick_spacing']))  # Plt ticks\n",
    "\n",
    "    # Axis labels\n",
    "    plt.title(temp_kwargs['title'], fontsize=12, fontname='Arial', color='k')\n",
    "    plt.ylabel(y_label, fontsize=12, fontname=\"Arial\", color='k', labelpad=5)\n",
    "    plt.xlabel('Position', fontsize=12,\n",
    "               fontname=\"Arial\", color='k', labelpad=0)\n",
    "\n",
    "    # x=y line\n",
    "    plt.plot([0, df['Position'].max()], [0, 1],\n",
    "             color='silver', lw=2, linestyle='--')\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _filter_bySNV(self, mode):\n",
    "\n",
    "    # Select all, SNV, nonSNV\n",
    "    if mode == 'all':\n",
    "        df = self.dataframe\n",
    "    elif mode == 'SNV':\n",
    "        df = self.dataframe_SNV\n",
    "    elif mode == 'nonSNV':\n",
    "        df = self.dataframe_nonSNV\n",
    "    df = df.groupby(by='Position', as_index=False).mean()\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Box Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.486292Z",
     "start_time": "2020-10-06T23:17:05.476473Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.666685Z",
     "iopub.status.busy": "2020-10-11T02:23:58.666360Z",
     "iopub.status.idle": "2020-10-11T02:23:58.676331Z",
     "shell.execute_reply": "2020-10-11T02:23:58.675762Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.666664Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_box(binned_x, y, output_file: Union[None, str, Path] = None,**kwargs):\n",
    "    '''\n",
    "    Generates a boxplot. Data needs to be binned prior before using this function. \n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    x, binned_y : arrays\n",
    "        Contain the data is going to plot\n",
    "        \n",
    "    **kwargs : other keyword arguments\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                \n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "\n",
    "    '''\n",
    "    # Load parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # Update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (2.5, 2))\n",
    "\n",
    "    # Make figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    \n",
    "    # Plot data\n",
    "    ax = sns.boxplot(binned_x, y, color='white', fliersize=2)\n",
    "\n",
    "    plt.setp(ax.artists, edgecolor='k', facecolor='w')\n",
    "    plt.setp(ax.lines, color='k')\n",
    "\n",
    "    # graph parameters\n",
    "    plt.title(temp_kwargs['title'], fontsize=10,\n",
    "              fontname='Arial', color='k', pad=8)\n",
    "    plt.ylabel(temp_kwargs['y_label'], fontsize=10,\n",
    "               fontname=\"Arial\", color='k', labelpad=0)\n",
    "    plt.xlabel(temp_kwargs['x_label'], fontsize=10,\n",
    "               fontname=\"Arial\", color='k')\n",
    "\n",
    "    # axes limits\n",
    "    plt.xlim(temp_kwargs['xscale'])\n",
    "    plt.ylim(temp_kwargs['yscale'])\n",
    "    plt.grid()\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    \n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "\n",
    "    return\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
