{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-11-25T19:59:27.839842Z",
     "iopub.status.busy": "2020-11-25T19:59:27.839538Z",
     "iopub.status.idle": "2020-11-25T19:59:29.083330Z",
     "shell.execute_reply": "2020-11-25T19:59:29.082072Z",
     "shell.execute_reply.started": "2020-11-25T19:59:27.839809Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate NNS primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-11-25T19:59:30.579534Z",
     "iopub.status.busy": "2020-11-25T19:59:30.579147Z",
     "iopub.status.idle": "2020-11-25T19:59:30.588129Z",
     "shell.execute_reply": "2020-11-25T19:59:30.587097Z",
     "shell.execute_reply.started": "2020-11-25T19:59:30.579508Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_primers(\n",
    "    dna,\n",
    "    start,\n",
    "    end,\n",
    "    output_file: Union[str, Path],\n",
    "    codon='NNS',\n",
    "    length_primer=15,\n",
    "    tm=None,\n",
    "    return_df=False\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate primers for saturation mutagenesis.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dna : string\n",
    "        DNA sequence containing the protein of study.\n",
    "        The DNA sequence should also contain at least 15 base pairs before the\n",
    "        starting ATG and 15 base pairs after the stop codon.\n",
    "\n",
    "    start : string\n",
    "        Beginning of the DNA sequence that will be mutageneized.\n",
    "        For example, if you will start mutating the first methionine, copy a few more dna bases\n",
    "        so the algorithm can identify it 'ATGACCAGC'.\n",
    "\n",
    "    end : string\n",
    "        The algorithm will stop creating primers once it reaches that base.\n",
    "        For example, if you will stop mutating at the stop codon, copy a few more dna bases ie. 'TAAATGATT'.\n",
    "\n",
    "    output_file : str\n",
    "        File where the list of primers will be exported to. Only exports to excel.\n",
    "        Example: 'path/primers.xlsx'.\n",
    "\n",
    "    codon : str, default 'NNS'\n",
    "        Degenerate codon that will be used to create the primers. Check idt's website for a list\n",
    "        of all mixed bases and letter code (https://www.idtdna.com/pages/products/custom-dna-rna/mixed-bases).\n",
    "        This parameter should contain 3 letters, although can contain more.\n",
    "\n",
    "    length_primer: int, default 15\n",
    "        Number of bases that the primers will have to each side of the mutated codon.\n",
    "        Total primer length will be 2*length_primer+3.\n",
    "\n",
    "    tm : int, default None\n",
    "        Melting temperature in Celsius of the primers. Will override length_primer.\n",
    "        If none, primers will have a total length of 2*length_primer+3\n",
    "\n",
    "    return_df : boolean, default False\n",
    "        If true, will export a dataframe with the primers.\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    df : pandas dataframe, optional\n",
    "        Dataframe containing the primers.\n",
    "\n",
    "    \"\"\"\n",
    "    # Transform to upper case\n",
    "    dna = dna.upper()\n",
    "\n",
    "    # Find the first and last codons\n",
    "    start_codon = dna.find(start.upper())\n",
    "    end_codon = dna.find(end.upper())\n",
    "\n",
    "    # loop through DNA and make a list with fp and second list with rp\n",
    "\n",
    "    label_fp = [\n",
    "        'fp ' + str(i) for i in range(0, int((end_codon - start_codon) / 3))\n",
    "    ]\n",
    "    label_rp = [\n",
    "        'rp ' + str(i) for i in range(0, int((end_codon - start_codon) / 3))\n",
    "    ]\n",
    "    forward_primers, reverse_primers = _create_primers_list(\n",
    "        dna, start_codon, end_codon, codon, length_primer, tm\n",
    "    )\n",
    "\n",
    "    # Create dataframe\n",
    "    dictionary = {\n",
    "        'FP_label': label_fp, 'FP_seq': forward_primers, 'RP_label': label_rp,\n",
    "        'RP_seq': reverse_primers\n",
    "    }\n",
    "    df = pd.DataFrame(dictionary)\n",
    "\n",
    "    # Export dataframe\n",
    "    if output_file:\n",
    "        df.to_excel(Path(output_file), sheet_name='Primers', index=False)\n",
    "\n",
    "    # Return dataframe\n",
    "    if return_df:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-11-25T19:59:31.432025Z",
     "iopub.status.busy": "2020-11-25T19:59:31.431822Z",
     "iopub.status.idle": "2020-11-25T19:59:31.439999Z",
     "shell.execute_reply": "2020-11-25T19:59:31.439026Z",
     "shell.execute_reply.started": "2020-11-25T19:59:31.432005Z"
    }
   },
   "outputs": [],
   "source": [
    "def _create_primers_list(dna, start_codon, end_codon, codon, length_primer, tm):\n",
    "    """Aux function to create list with fp and list with rp"""\n",
    "    forward_primers = []\n",
    "    reverse_primers = []\n",
    "    for codonposition in range(start_codon, end_codon, 3):\n",
    "        # Create fp, rp for that position\n",
    "        fp, rp = _primerdesign(dna, codon, codonposition, length_primer, tm)\n",
    "        # Append to list\n",
    "        forward_primers.append(fp)\n",
    "        reverse_primers.append(rp)\n",
    "    return forward_primers, reverse_primers\n",
    "\n",
    "\n",
    "def _reverse_complement(dna):\n",
    "    """aux function that uses biopython to calculate the reverse complement of a DNA string.\n",
    "    Includes mixed-base code. More info in https://biopython.org/docs/1.75/api/Bio.Seq.html"""\n",
    "\n",
    "    # Needs to be converted to str\n",
    "    reverse_dna = str(Seq(dna).reverse_complement())\n",
    "\n",
    "    return reverse_dna\n",
    "\n",
    "\n",
    "def _primerdesign(dna, codon, codonposition, length_primer, tm):\n",
    "    """aux function to design the degenerate primers given a sequence and a codon position. \n",
    "    The length of the primer is fixed.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dna : string\n",
    "        DNA sequence containing the protein of study. \n",
    "        The DNA sequence should also contain at least 15 base pairs before the \n",
    "        starting ATG and 15 base pairs after the stop codon.\n",
    "\n",
    "    codon : str\n",
    "        Degenerate codon that will be used to create the primers. Check idt's website for a list\n",
    "        of all mixed bases and letter code (https://www.idtdna.com/pages/products/custom-dna-rna/mixed-bases).\n",
    "\n",
    "    codonposition : int\n",
    "        Position of the codon  to mutate with respect to the gene. \n",
    "        The first codon is 0 and the mutation occurs after the inputed codon number.\n",
    "\n",
    "    length_primer : int, default 15\n",
    "        Number of bases that the primers will have to each side of the mutated codon.\n",
    "        Total primer length will be 2*length_primer+3.\n",
    "        If using tm to generate primer length, then set length_primer = None. \n",
    "\n",
    "    tm : int, default None\n",
    "        Melting temperature in Celsius of the primers. Will override length_primer.\n",
    "        If none, primers will have a total length of 2*length_primer+3\n",
    "        \n",
    "    Returns\n",
    "    ---------\n",
    "    forward_primer, reverse_primer\n",
    "    """\n",
    "\n",
    "    if tm:\n",
    "        # loop until tm is achieved\n",
    "        x = meltingT_fp = 6\n",
    "        while meltingT_fp < tm:\n",
    "            forward_primer = dna[(codonposition-x):codonposition] + \\\n",
    "                codon + dna[(codonposition+3):(codonposition+x)]\n",
    "            meltingT_fp = mt.Tm_NN(forward_primer)\n",
    "            x += 1\n",
    "    else:\n",
    "        forward_primer = dna[(codonposition-length_primer):codonposition] + \\\n",
    "            codon + dna[(codonposition+3):(codonposition+length_primer+3)]\n",
    "\n",
    "    reverse_primer = _reverse_complement(forward_primer)\n",
    "    return forward_primer, reverse_primer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate Variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-11-25T19:59:41.848904Z",
     "iopub.status.busy": "2020-11-25T19:59:41.848685Z",
     "iopub.status.idle": "2020-11-25T19:59:41.861113Z",
     "shell.execute_reply": "2020-11-25T19:59:41.860190Z",
     "shell.execute_reply.started": "2020-11-25T19:59:41.848883Z"
    }
   },
   "outputs": [],
   "source": [
    "def create_variants(\n",
    "    dna, codon_list, output_file: Union[str, Path], return_df=False\n",
    "):\n",
    "    """\n",
    "    Generate a list of all point mutants given a dna sequence and a list of codons.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dna : str, \n",
    "        Contains the DNA sequence of the allele of reference (usually wild-type).\n",
    "\n",
    "    codon_list : list or str\n",
    "        Input a list of the codons that were used to create point mutations. Example: [\"GCC\", \"GCG\", \"TGC\"].\n",
    "        It is important to know that the order of the codon_list will determine the output order.\n",
    "\n",
    "    output_file : str\n",
    "        File where the list of primers will be exported to. Only exports to 'xlsx',\n",
    "        'fasta', 'txt'.\n",
    "        Example: 'path/sequences.xlsx'. \n",
    "\n",
    "    return_df : boolean, default False\n",
    "        If true, will export a dataframe with the primers.\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    df : pandas dataframe, optional\n",
    "        Dataframe containing the generated sequences.\n",
    "\n",
    "    """\n",
    "    # Make upper case in case input was lower case\n",
    "    dna = dna.upper()\n",
    "    codon_list = [item.upper() for item in codon_list]\n",
    "\n",
    "    # Generate list of variants\n",
    "    seq_list = _enumerate_variants_2(dna, codon_list)\n",
    "\n",
    "    # Make dataframe\n",
    "    df = pd.DataFrame()\n",
    "    df['Sequences'] = seq_list\n",
    "\n",
    "    # Export to excel or fasta\n",
    "    if output_file:\n",
    "        if Path(output_file).suffix == '.xlsx':\n",
    "            df.to_excel(Path(output_file), sheet_name='Variants', index=False)\n",
    "        elif Path(output_file).suffix == '.fasta' or Path(output_file\n",
    "                                                          ).suffix == '.txt':\n",
    "            _list_to_fasta(seq_list, output_file)\n",
    "\n",
    "    # Return dataframe\n",
    "    if return_df:\n",
    "        return df\n",
    "\n",
    "\n",
    "def _enumerate_variants_2(dna, codon_list):\n",
    "    """\n",
    "    Copy of _enumerate variants with slight changes. Does return the wild-type sequence\n",
    "    as the first item of the list.\n",
    "    """\n",
    "    # Create list with codons of sequence\n",
    "    wtSeqList = [dna[i:i + 3] for i in range(0, len(dna), 3)]\n",
    "\n",
    "    # List of sequences\n",
    "    seq_list = [dna]\n",
    "\n",
    "    # Loop over the dna sequence\n",
    "    for position in range(0, len(wtSeqList)):\n",
    "        for codon in (codon_list):\n",
    "            variant = ''.join(wtSeqList[0:position]) + \\\n",
    "                codon + ''.join(wtSeqList[position+1:])\n",
    "            if (variant != dna):\n",
    "                seq_list.append(variant)\n",
    "    return seq_list\n",
    "\n",
    "\n",
    "def _list_to_fasta(seq_list, output_file):\n",
    "    """Export list to fasta format"""\n",
    "\n",
    "    # Open file\n",
    "    ofile = open(str(output_file), \"w\")\n",
    "\n",
    "    # Loop through list and write into file\n",
    "    for i, seq in enumerate(seq_list):\n",
    "        line = (\">{}\\n{}\\n\").format(str(i), seq)\n",
    "        ofile.write(line)\n",
    "\n",
    "    # Close file\n",
    "    ofile.close()\n",
    "    return"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
