{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    },
    "execution": {
     "iopub.execute_input": "2021-02-17T03:23:21.346335Z",
     "iopub.status.busy": "2021-02-17T03:23:21.346126Z",
     "iopub.status.idle": "2021-02-17T03:23:22.141324Z",
     "shell.execute_reply": "2021-02-17T03:23:22.140730Z",
     "shell.execute_reply.started": "2021-02-17T03:23:21.346314Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from code_kwargs.ipynb\n",
      "importing Jupyter notebook from code_utils.ipynb\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import copy\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.ticker as ticker\n",
    "from os import path\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "from collections import defaultdict, Counter\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "# local modules\n",
    "try:\n",
    "    import mutagenesis_visualization.main.scripts.code_kwargs as code_kwargs\n",
    "    import mutagenesis_visualization.main.scripts.code_utils as code_utils\n",
    "except ModuleNotFoundError:\n",
    "    import import_notebook\n",
    "    import code_kwargs\n",
    "    import code_utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heatmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-02-17T03:31:11.092026Z",
     "iopub.status.busy": "2021-02-17T03:31:11.091698Z",
     "iopub.status.idle": "2021-02-17T03:31:11.121862Z",
     "shell.execute_reply": "2021-02-17T03:31:11.121144Z",
     "shell.execute_reply.started": "2021-02-17T03:31:11.091993Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_heatmap(\n",
    "    self,\n",
    "    nancolor='lime',\n",
    "    show_cartoon=False,\n",
    "    show_snv=False,\n",
    "    hierarchical=False,\n",
    "    output_file: Union[None, str, Path] = None,\n",
    "    **kwargs\n",
    "):\n",
    "    '''\n",
    "    Generate a heatmap plot of the enrichment scores.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    nancolor : str, default 'lime'\n",
    "        Will color np.nan values with the specified color.\n",
    "    \n",
    "    show_carton : boolean, default False\n",
    "        If true, the plot will display a cartoon with the secondary structure. \n",
    "        The user must have added the secondary structure to the object. \n",
    "    \n",
    "    show_snv : boolean, default False\n",
    "        If true, it will only display mutants that are a single nucleotide variant (SNV) of the wild-type\n",
    "        protein sequence. The algorithm does not take into account the wild-type DNA allele, so it \n",
    "        will include any possible mutant that is one base away.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                \n",
    "    **kwargs : other keyword arguments\n",
    "        return_plot_object : boolean, default False\n",
    "            If true, will return plotting objects (ie. fig, ax).        \n",
    "    Returns\n",
    "    ----------\n",
    "    fig, ax, ax2, ax3, averageresidue : matplotlib figure and subplots\n",
    "        Needs to have return_plot_object==True. By default they do\n",
    "        not get returned.\n",
    "        \n",
    "    '''\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "\n",
    "    # load labels\n",
    "    temp_kwargs['color_sequencelabels'] = _labels(self.start_position)[0]\n",
    "    temp_kwargs['number_sequencelabels'] = _labels(self.start_position)[1]\n",
    "\n",
    "    # sort data by rows in specified order by user\n",
    "    df = code_utils._df_rearrange(\n",
    "        code_utils._add_SNV_boolean(self.dataframe_stopcodons.copy()),\n",
    "        temp_kwargs['neworder_aminoacids'],\n",
    "        values='Score_NaN',\n",
    "        show_snv=show_snv\n",
    "    )\n",
    "\n",
    "    # average of residues by positon\n",
    "    average = [\n",
    "        code_utils._add_SNV_boolean(self.dataframe.copy()\n",
    "                                    ).groupby(by='Position').mean()['Score_NaN']\n",
    "    ]\n",
    "    # For 1 column case\n",
    "    #return (average)\n",
    "    if len(average) == 1: \n",
    "        average = [np.array(average[0])]\n",
    "    \n",
    "    # Create new sequence that we may to change order later\n",
    "    self.sequence_updated = self.sequence\n",
    "\n",
    "    # sort data by columns if specified by user\n",
    "    if hierarchical:\n",
    "        sorted_columns = _hierarchical_sort(df)\n",
    "        sorted_columns_corrected = sorted_columns + self.start_position  # adds the initial number so index match\n",
    "        df = df[sorted_columns_corrected]\n",
    "        temp = pd.DataFrame(\n",
    "            average\n",
    "        )  # had to convert back to df to be able to sort properly\n",
    "        average = temp[sorted_columns_corrected]\n",
    "        # Change the sequence order\n",
    "        self.sequence_updated = pd.DataFrame(list(self.sequence)\n",
    "                                             ).T[sorted_columns]\n",
    "        self.sequence_updated = list(self.sequence_updated.iloc[0])\n",
    "\n",
    "    # declare figure and subplots\n",
    "    figwidth = 14 * len(df.columns) / 165\n",
    "    temp_kwargs['figsize_x'] = kwargs.get('figsize_x', figwidth)\n",
    "    # Change parameters depending on whether cartoon is on or off\n",
    "    if show_cartoon:\n",
    "        figheight = 2.45\n",
    "        temp_kwargs['figsize_y'] = kwargs.get('figsize_y', figheight)\n",
    "        fig = plt.figure(figsize=(temp_kwargs['figsize_x'], temp_kwargs['figsize_y']))\n",
    "        gs = gridspec.GridSpec(\n",
    "            nrows=3,\n",
    "            ncols=2,\n",
    "            height_ratios=[len(df), 1, 5],\n",
    "            width_ratios=[len(df.columns), 1]\n",
    "        )\n",
    "    else:\n",
    "        figheight = 2\n",
    "        temp_kwargs['figsize_y'] = kwargs.get('figsize_y', figheight)\n",
    "        fig = plt.figure(figsize=(temp_kwargs['figsize_x'], temp_kwargs['figsize_y']))\n",
    "        gs = gridspec.GridSpec(\n",
    "            nrows=2,\n",
    "            ncols=2,\n",
    "            height_ratios=[len(df), 1],\n",
    "            width_ratios=[len(df.columns), 1]\n",
    "        )\n",
    "\n",
    "    ax = plt.subplot(gs[0, 0])\n",
    "    averageresidue = plt.subplot(gs[1, 0])\n",
    "    cbar1 = plt.subplot(gs[0, 1])\n",
    "    cbar2 = plt.subplot(gs[1, 1])\n",
    "\n",
    "    # Change color of values that are NaN\n",
    "    cmap = temp_kwargs['colormap']\n",
    "    cmap.set_bad(color=nancolor)\n",
    "\n",
    "    # main heatmap\n",
    "    heatmap = ax.pcolormesh(\n",
    "        df,\n",
    "        vmin=temp_kwargs['colorbar_scale'][0],\n",
    "        vmax=temp_kwargs['colorbar_scale'][1],\n",
    "        cmap=cmap,\n",
    "        edgecolors='k',\n",
    "        linewidths=0.2,\n",
    "        antialiased=True,\n",
    "        color='darkgrey'\n",
    "    )\n",
    "\n",
    "    # average by position\n",
    "    heatmapaverageresidues = averageresidue.pcolormesh(\n",
    "        average,\n",
    "        vmin=temp_kwargs['colorbar_scale'][0],\n",
    "        vmax=temp_kwargs['colorbar_scale'][1],\n",
    "        cmap=cmap,\n",
    "        edgecolors='k',\n",
    "        linewidths=0.2,\n",
    "        antialiased=True,\n",
    "        color='darkgrey'\n",
    "    )\n",
    "\n",
    "    # ____________axes manipulation____________________________________________\n",
    "    # put the major ticks at the middle of each cell\n",
    "    ax.set_xticks(np.arange(len(df.columns)) + 0.5, minor=False)\n",
    "    ax.set_yticks(np.arange(len(df)) + 0.5, minor=False)\n",
    "\n",
    "    # position of axis labels\n",
    "    ax.tick_params('x', direction='out', pad=-2.5)\n",
    "    ax.tick_params('y', direction='out', pad=0.4)\n",
    "\n",
    "    # make new axes\n",
    "    ax2 = ax.twiny()\n",
    "    ax3 = ax.twinx()\n",
    "\n",
    "    # tune the axes\n",
    "    ax2.set_xticks(np.arange(len(df.columns)) + 0.5, minor=False)\n",
    "    ax3.set_yticks(np.arange(len(df)) + 0.5, minor=False)\n",
    "    ax2.tick_params(direction='out', pad=4)\n",
    "    ax3.tick_params(direction='out', pad=0.4)\n",
    "    averageresidue.tick_params(direction='out', pad=-2)\n",
    "    averageresidue.set_xticks(\n",
    "        np.arange(len(df.columns)) + 0.5,\n",
    "        minor=False,\n",
    "    )\n",
    "    averageresidue.set_yticks(np.arange(0.5) + 0.5)\n",
    "\n",
    "    # Set the limits of the new axis from the original axis limits\n",
    "    ax2.set_xlim(ax.get_xlim())\n",
    "    ax3.set_ylim(ax.get_ylim())\n",
    "\n",
    "    # want a more natural, table-like display\n",
    "    ax.invert_yaxis()\n",
    "    ax.xaxis.tick_top()\n",
    "    ax3.invert_yaxis()\n",
    "\n",
    "    # remove ticks\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "    ax2.yaxis.set_ticks_position('none')\n",
    "    ax2.xaxis.set_ticks_position('none')\n",
    "    ax3.yaxis.set_ticks_position('none')\n",
    "    averageresidue.xaxis.set_ticks_position('none')\n",
    "    averageresidue.yaxis.set_ticks_position('none')\n",
    "\n",
    "    # so labels of x and y do not show up and my labels show up instead\n",
    "    ax.set_xticklabels(\n",
    "        list(self.sequence_updated),\n",
    "        fontsize=6.5,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    ax.set_yticklabels(\n",
    "        temp_kwargs['neworder_aminoacids'],\n",
    "        fontsize=6,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    # For numbering labels, change if hierarchical sorting is true\n",
    "    if not (hierarchical):\n",
    "        ax2.set_xticklabels(\n",
    "            temp_kwargs['number_sequencelabels'][0:len(df.columns)],\n",
    "            fontsize=10,\n",
    "            fontname=\"Arial\",\n",
    "            color='k',\n",
    "            minor=False\n",
    "        )\n",
    "    else:\n",
    "        ax2.tick_params(direction='out', pad=7)\n",
    "        ax2.set_xticklabels(\n",
    "            sorted_columns_corrected,\n",
    "            fontsize=5,\n",
    "            fontname=\"Arial\",\n",
    "            color='k',\n",
    "            minor=False,\n",
    "            rotation=90\n",
    "        )\n",
    "    ax3.set_yticklabels(\n",
    "        temp_kwargs['neworder_aminoacids'],\n",
    "        fontsize=6,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    averageresidue.set_xticklabels(\n",
    "        list(self.sequence_updated),\n",
    "        fontsize=6.5,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    rowaverage = ''\n",
    "    averageresidue.set_yticklabels(\n",
    "        rowaverage, fontsize=6, fontname=\"Arial\", color='k', minor=False\n",
    "    )\n",
    "\n",
    "    # align the labels of the y axis\n",
    "    for ylabel in ax.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "    for ylabel in ax3.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "\n",
    "    # for coloring the residues that are 10,20...\n",
    "    for xtick, color in zip(ax.get_xticklabels(),\n",
    "                            temp_kwargs['color_sequencelabels']):\n",
    "        xtick.set_color(color)\n",
    "    for xtick, color in zip(averageresidue.get_xticklabels(),\n",
    "                            temp_kwargs['color_sequencelabels']):\n",
    "        xtick.set_color(color)\n",
    "    # _____________________________________________________________________________\n",
    "\n",
    "    # for color bar format\n",
    "    cbar1.axis('off')\n",
    "    cbar2.axis('off')\n",
    "    cb = plt.colorbar(\n",
    "        heatmap,\n",
    "        fraction=1,\n",
    "        pad=0,\n",
    "        ax=[cbar1],\n",
    "        aspect=5,\n",
    "        ticks=[\n",
    "            temp_kwargs['colorbar_scale'][0],\n",
    "            np.mean(temp_kwargs['colorbar_scale']),\n",
    "            temp_kwargs['colorbar_scale'][1]\n",
    "        ],\n",
    "        orientation='vertical'\n",
    "    )\n",
    "    cb.ax.set_yticklabels(\n",
    "        cb.ax.get_yticklabels(), fontsize=6, fontname=\"Arial\", color='k'\n",
    "    )\n",
    "    cb.update_ticks()\n",
    "    plt.text(\n",
    "        1.2 + 10 / len(df.columns),\n",
    "        0.7,\n",
    "        r'$\\langle∆E^x_i\\rangle_x$',\n",
    "        transform=cbar1.transAxes,\n",
    "        horizontalalignment='center',\n",
    "        fontsize=7,\n",
    "        fontname=\"Arial\",\n",
    "        color='k'\n",
    "    )\n",
    "\n",
    "    gs.update(hspace=0.1, wspace=0.1 / len(df.columns) * 50)\n",
    "\n",
    "    # for putting title on graph\n",
    "    plt.title(\n",
    "        temp_kwargs['title'],\n",
    "        horizontalalignment='center',\n",
    "        fontname=\"Arial\",\n",
    "        fontsize=12\n",
    "    )\n",
    "\n",
    "    # Cartoon\n",
    "    if show_cartoon:\n",
    "        _generate_cartoon(self, gs, 2, temp_kwargs['cartoon_colors'], 0.025)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    # return matplotlib object\n",
    "    if temp_kwargs['return_plot_object']:\n",
    "        return fig, ax, ax2, ax3, averageresidue\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:04.900609Z",
     "start_time": "2020-10-06T23:17:04.823223Z"
    },
    "execution": {
     "iopub.execute_input": "2021-02-17T03:23:29.359372Z",
     "iopub.status.busy": "2021-02-17T03:23:29.358874Z",
     "iopub.status.idle": "2021-02-17T03:23:29.393531Z",
     "shell.execute_reply": "2021-02-17T03:23:29.392812Z",
     "shell.execute_reply.started": "2021-02-17T03:23:29.359339Z"
    }
   },
   "outputs": [],
   "source": [
    "def _labels(start_position=1):\n",
    "    # residue label and color\n",
    "    emptylist = [''] * 1000\n",
    "    number_sequencelabels = list([\n",
    "        'b' if index in np.arange(10 - (start_position % 10), 1000, 10) else 'k'\n",
    "        for index, x in enumerate(emptylist)\n",
    "    ])\n",
    "    color_sequencelabels = list([\n",
    "        index + start_position\n",
    "        if index in np.arange(10 - (start_position % 10), 1000, 10) else ''\n",
    "        for index, x in enumerate(emptylist)\n",
    "    ])\n",
    "    return number_sequencelabels, color_sequencelabels\n",
    "\n",
    "\n",
    "def _generate_cartoon(\n",
    "    self,\n",
    "    gs,\n",
    "    n_row,\n",
    "    colors,\n",
    "    bottom_space=0,\n",
    "    fig_inches=13.91,\n",
    "    show_labels=True\n",
    "):\n",
    "    '''Generates cartoon for heatmap'''\n",
    "    # Create subplot\n",
    "    cartoon = plt.subplot(gs[n_row, 0])\n",
    "\n",
    "    # Generate coordinates of labels\n",
    "    labels = list(Counter(self.secondary).keys())\n",
    "    length = list(Counter(self.secondary).values())\n",
    "    cumsum = length[:-1]\n",
    "    cumsum.insert(0, self.start_position)\n",
    "    cumsum = np.cumsum(cumsum)\n",
    "\n",
    "    # Create cartoon\n",
    "    for label, length, cum in zip(labels, length, cumsum):\n",
    "        if 'β' in label:\n",
    "            loopstructure = _loop(cum, length, color=colors[2])\n",
    "            cartoon.add_patch(loopstructure)\n",
    "            sheetstructure = _sheet(cum, length, colors[0])\n",
    "            cartoon.add_patch(sheetstructure)\n",
    "            x_label = cum + length - 3.5\n",
    "            if length > 2 and show_labels:  # If beta sheet is too small, label does not fit\n",
    "                if length == 3:\n",
    "                    cartoon.text((x_label + 0.6),\n",
    "                                 -0.25,\n",
    "                                 label,\n",
    "                                 name='Arial',\n",
    "                                 fontweight='normal',\n",
    "                                 size=8.5 * fig_inches / 13.91,\n",
    "                                 multialignment='right')\n",
    "                else:\n",
    "                    cartoon.text((x_label),\n",
    "                                 -0.25,\n",
    "                                 label,\n",
    "                                 name='Arial',\n",
    "                                 fontweight='normal',\n",
    "                                 size=8.5 * fig_inches / 13.91,\n",
    "                                 multialignment='right')\n",
    "        elif 'α' in label:\n",
    "            helixstructure = _helix(cum, length, colors[1])\n",
    "            cartoon.add_patch(helixstructure)\n",
    "            x_label = cum + length / 2 - 1\n",
    "            if length > 2 and show_labels:\n",
    "                cartoon.text((x_label),\n",
    "                             -0.3,\n",
    "                             label,\n",
    "                             name='Arial',\n",
    "                             fontweight='normal',\n",
    "                             size=9 * fig_inches / 14,\n",
    "                             multialignment='center')\n",
    "        elif 'L' in label:\n",
    "            loopstructure = _loop(cum, length, colors[2])\n",
    "            cartoon.add_patch(loopstructure)\n",
    "\n",
    "    # format of secondary cartoon\n",
    "    cartoon.xaxis.set_ticks_position('none')\n",
    "    cartoon.yaxis.set_ticks_position('none')\n",
    "    cartoon.axis('off')\n",
    "\n",
    "    # size\n",
    "    cartoon.set_xlim(\n",
    "        self.start_position - 0.1,\n",
    "        len(self.secondary) + self.start_position + 0.2\n",
    "    )\n",
    "    cartoon.set_ylim(-2, 2.5)\n",
    "\n",
    "    # adjust proximity to heatmap\n",
    "    box = cartoon.get_position()\n",
    "    box.y0 = box.y0 - bottom_space\n",
    "    box.y1 = box.y1 - bottom_space\n",
    "    cartoon.set_position(box)\n",
    "\n",
    "    return\n",
    "\n",
    "\n",
    "def _sheet(starting_aa, length_aa, color='lightgreen'):\n",
    "    dx = length_aa\n",
    "    sheetstructure = patches.FancyArrow(\n",
    "        starting_aa,\n",
    "        0.25,\n",
    "        dx,\n",
    "        0,\n",
    "        width=2,\n",
    "        length_includes_head=True,\n",
    "        head_width=4,\n",
    "        head_length=3,\n",
    "        shape='full',\n",
    "        overhang=0,\n",
    "        head_starts_at_zero=False,\n",
    "        ec='k',\n",
    "        fc=color\n",
    "    )\n",
    "    return sheetstructure\n",
    "\n",
    "\n",
    "def _helix(starting_aa, length_aa, color='lavender'):\n",
    "    \"\"\"produces matplotlib.Rectangle of length specified by length_aa \"\"\"\n",
    "    dx = length_aa  # so i can overlap tip\n",
    "    helixstructure = plt.Rectangle((starting_aa, -0.85),\n",
    "                                   dx,\n",
    "                                   2.2,\n",
    "                                   fc=color,\n",
    "                                   ec='k')\n",
    "    return helixstructure\n",
    "\n",
    "\n",
    "def _loop(starting_aa, length_aa, color='k'):\n",
    "    dx = length_aa\n",
    "    loopstructure = plt.Rectangle((starting_aa, 0), dx, 0.5, fc=color)\n",
    "    return loopstructure\n",
    "\n",
    "\n",
    "def _hierarchical_sort(df):\n",
    "    \"\"\"\n",
    "    sorts columns of dataset using hierarchical clustering\n",
    "    returns order to rearrange dataset\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    # replaces NaN values with 0\n",
    "    df_1 = df.fillna(0)\n",
    "\n",
    "    # give sorted column order\n",
    "    Z = hierarchy.ward(df_1.T)\n",
    "    new_order = hierarchy.leaves_list(Z)\n",
    "\n",
    "    return new_order"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grouped Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:04.929321Z",
     "start_time": "2020-10-06T23:17:04.904408Z"
    },
    "execution": {
     "iopub.execute_input": "2021-02-17T03:23:29.555881Z",
     "iopub.status.busy": "2021-02-17T03:23:29.551075Z",
     "iopub.status.idle": "2021-02-17T03:23:29.594347Z",
     "shell.execute_reply": "2021-02-17T03:23:29.593682Z",
     "shell.execute_reply.started": "2021-02-17T03:23:29.555820Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_heatmap_rows(\n",
    "    self,\n",
    "    selection=['E', 'Q', 'A', 'P', 'V', 'Y'],\n",
    "    nancolor='lime',\n",
    "    output_file: Union[None, str, Path] = None,\n",
    "    **kwargs\n",
    "):\n",
    "    '''\n",
    "    Generate a heatmap plot enrichment scores of selected aminoacids. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    selection : list of aa to show, default ['E','Q','A','P','V','Y']. \n",
    "    \n",
    "    nancolor : str, default 'lime'\n",
    "        Will color np.nan values with the specified color.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "        return_plot_object : boolean, default False\n",
    "            If true, will return plotting objects (ie. fig, ax).\n",
    "    Returns\n",
    "    ----------\n",
    "    fig, ax, ax2 : matplotlib figure and subplots\n",
    "        Needs to have return_plot_object==True. By default they do\n",
    "        not get returned.\n",
    "        \n",
    "    '''\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "\n",
    "    # load labels\n",
    "    temp_kwargs['color_sequencelabels'] = _labels(self.start_position)[0]\n",
    "    temp_kwargs['number_sequencelabels'] = _labels(self.start_position)[1]\n",
    "\n",
    "    # Add group and pivot df\n",
    "    df = code_utils._select_aa(\n",
    "        self.dataframe_stopcodons, selection, values='Score_NaN'\n",
    "    )\n",
    "    dataset = df.to_numpy()\n",
    "\n",
    "    # The size can be changed. I found it empirically\n",
    "    figwidth = 14 * len(dataset[0]) / 165\n",
    "    figheight = 2 / 21 * len(selection)\n",
    "    temp_kwargs['figsize_x'] = kwargs.get('figsize_x', figwidth)\n",
    "    temp_kwargs['figsize_y'] = kwargs.get('figsize_y', figheight)\n",
    "\n",
    "    fig = plt.figure(figsize=(temp_kwargs['figsize_x'], temp_kwargs['figsize_y']))\n",
    "    gs = gridspec.GridSpec(nrows=1, ncols=2, width_ratios=[len(dataset[0]), 1])\n",
    "    ax = plt.subplot(gs[0, 0])\n",
    "    cbar1 = plt.subplot(gs[0, 1])\n",
    "\n",
    "    # Change color of values that are NaN\n",
    "    cmap = temp_kwargs['colormap']\n",
    "    cmap.set_bad(color=nancolor)\n",
    "\n",
    "    # main heatmap\n",
    "    heatmap = ax.pcolormesh(\n",
    "        dataset,\n",
    "        vmin=temp_kwargs['colorbar_scale'][0],\n",
    "        vmax=temp_kwargs['colorbar_scale'][1],\n",
    "        cmap=cmap,\n",
    "        edgecolors='k',\n",
    "        linewidths=0.2,\n",
    "        antialiased=True,\n",
    "        color='darkgrey'\n",
    "    )\n",
    "\n",
    "    # put the major ticks at the middle of each cell\n",
    "    ax.set_xticks(np.arange(dataset.shape[1]) + 0.5, minor=False)\n",
    "    ax.set_yticks(np.arange(dataset.shape[0]) + 0.5, minor=False)\n",
    "\n",
    "    # position of axis labels\n",
    "    ax.tick_params('x', direction='out', pad=-2.5)\n",
    "    ax.tick_params('y', direction='out', pad=0.4)\n",
    "\n",
    "    # second axis\n",
    "    ax2 = ax.twiny()\n",
    "    ax2.set_xticks(np.arange(dataset.shape[1]) + 0.5, minor=False)\n",
    "    ax2.tick_params(direction='out', pad=4)\n",
    "\n",
    "    # Set the limits of the new axis from the original axis limits\n",
    "    ax2.set_xlim(ax.get_xlim())\n",
    "\n",
    "    # want a more natural, table-like display\n",
    "    ax.invert_yaxis()\n",
    "    ax.xaxis.tick_top()\n",
    "\n",
    "    # so labels of x and y do not show up and my labels show up instead\n",
    "    ax.set_xticklabels(\n",
    "        list(self.sequence),\n",
    "        fontsize=6.5,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    ax.set_yticklabels(\n",
    "        list(df.T.columns),\n",
    "        fontsize=6,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    ax2.set_xticklabels(\n",
    "        temp_kwargs['number_sequencelabels'][0:len(dataset[0])],\n",
    "        fontsize=10,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "\n",
    "    # align the labels of the y axis\n",
    "    for ylabel in ax.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "\n",
    "    # for coloring the residues that are 10,20...\n",
    "    for xtick, color in zip(ax.get_xticklabels(),\n",
    "                            temp_kwargs['color_sequencelabels']):\n",
    "        xtick.set_color(color)\n",
    "\n",
    "    # for putting title on graph\n",
    "    plt.title(\n",
    "        temp_kwargs['title'],\n",
    "        horizontalalignment='center',\n",
    "        fontname=\"Arial\",\n",
    "        fontsize=12\n",
    "    )\n",
    "\n",
    "    # for color bar format\n",
    "    cbar1.axis('off')\n",
    "    cb = plt.colorbar(\n",
    "        heatmap,\n",
    "        fraction=1,\n",
    "        pad=0,\n",
    "        ax=[cbar1],\n",
    "        aspect=5,\n",
    "        ticks=[\n",
    "            temp_kwargs['colorbar_scale'][0],\n",
    "            np.mean(temp_kwargs['colorbar_scale']),\n",
    "            temp_kwargs['colorbar_scale'][1]\n",
    "        ],\n",
    "        orientation='vertical'\n",
    "    )\n",
    "    cb.ax.set_yticklabels(\n",
    "        cb.ax.get_yticklabels(), fontsize=8, fontname=\"Arial\", color='k'\n",
    "    )\n",
    "    cb.update_ticks()\n",
    "    gs.update(hspace=0.1, wspace=0.1 / len(dataset[0]) * 50)\n",
    "\n",
    "    # remove ticks\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "    ax2.yaxis.set_ticks_position('none')\n",
    "    ax2.xaxis.set_ticks_position('none')\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    # return matplotlib object\n",
    "    if temp_kwargs['return_plot_object']:\n",
    "        return fig, ax, ax2\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:04.948052Z",
     "start_time": "2020-10-06T23:17:04.931998Z"
    },
    "execution": {
     "iopub.execute_input": "2021-02-17T03:23:29.710875Z",
     "iopub.status.busy": "2021-02-17T03:23:29.710431Z",
     "iopub.status.idle": "2021-02-17T03:23:29.735840Z",
     "shell.execute_reply": "2021-02-17T03:23:29.733611Z",
     "shell.execute_reply.started": "2021-02-17T03:23:29.710814Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_heatmap_columns(\n",
    "    self,\n",
    "    segment,\n",
    "    ylabel_color='k',\n",
    "    nancolor='lime',\n",
    "    output_file: Union[None, str, Path] = None,\n",
    "    **kwargs\n",
    "):\n",
    "    '''\n",
    "    Generate a heatmap plot enrichment scores but only plots a selected segment.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    segment : list\n",
    "        Segment is typed as [20,40] and includes both residues 20 and 40.\n",
    "    \n",
    "    ylabel_color : str, default 'k'\n",
    "        Choose white if you don't want amino acid y axis label.\n",
    "    \n",
    "    nancolor : str, default 'lime'\n",
    "        Will color np.nan values with the specified color.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "        return_plot_object : boolean, default False\n",
    "            If true, will return plotting objects (ie. fig, ax).\n",
    "        \n",
    "    Returns\n",
    "    ----------\n",
    "    fig, ax, ax2 : matplotlib figure and subplots\n",
    "        Needs to have return_plot_object==True. By default they do\n",
    "        not get returned.\n",
    "        \n",
    "    '''\n",
    "\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    \n",
    "    \n",
    "    # load labels\n",
    "    temp_kwargs['color_sequencelabels'] = _labels(self.start_position)[0]\n",
    "    temp_kwargs['number_sequencelabels'] = _labels(self.start_position)[1]\n",
    "\n",
    "    # sort data in specified order by user\n",
    "    df_whole = code_utils._df_rearrange(\n",
    "        self.dataframe_stopcodons,\n",
    "        temp_kwargs['neworder_aminoacids'],\n",
    "        values='Score_NaN'\n",
    "    )\n",
    "\n",
    "    # select subset\n",
    "    c0 = segment[0] - self.start_position\n",
    "    c1 = segment[1] - self.start_position + 1\n",
    "    df = df_whole.iloc[:, c0:c1]\n",
    "\n",
    "    # the size can be changed\n",
    "    figwidth = 2 * len(df.columns) / 22\n",
    "    figheight = 2\n",
    "    temp_kwargs['figsize_x'] = kwargs.get('figsize_x', figwidth)\n",
    "    temp_kwargs['figsize_y'] = kwargs.get('figsize_y', figheight)\n",
    "\n",
    "    fig = plt.figure(figsize=(temp_kwargs['figsize_x'], temp_kwargs['figsize_y']))\n",
    "    gs = gridspec.GridSpec(nrows=1, ncols=1)\n",
    "    # needed to set autoscale off to avoid missalignment\n",
    "    ax = plt.subplot(gs[0])\n",
    "\n",
    "    # Change color of values that are NaN\n",
    "    cmap = temp_kwargs['colormap']\n",
    "    cmap.set_bad(color=nancolor)\n",
    "\n",
    "    # main heatmap\n",
    "    heatmap = ax.pcolormesh(\n",
    "        df,\n",
    "        vmin=temp_kwargs['colorbar_scale'][0],\n",
    "        vmax=temp_kwargs['colorbar_scale'][1],\n",
    "        cmap=cmap,\n",
    "        edgecolors='k',\n",
    "        linewidths=0.2,\n",
    "        antialiased=True,\n",
    "        color='darkgrey'\n",
    "    )\n",
    "\n",
    "    # put the major ticks at the middle of each cell\n",
    "    ax.set_xticks(\n",
    "        np.arange(len(df.columns)) + 0.5,\n",
    "        minor=False,\n",
    "    )\n",
    "    ax.set_yticks(np.arange(len(df)) + 0.5, minor=False)\n",
    "\n",
    "    # position of axis labels\n",
    "    ax.tick_params('x', direction='out', pad=-2.5)\n",
    "    ax.tick_params('y', direction='out', pad=0.4)\n",
    "\n",
    "    # second axis\n",
    "    ax2 = ax.twiny()\n",
    "    ax2.set_xticks(np.arange(len(df.columns)) + 0.5, minor=False)\n",
    "    ax2.tick_params(direction='out', pad=4)\n",
    "\n",
    "    # Set the limits of the new axis from the original axis limits\n",
    "    ax2.set_xlim(ax.get_xlim())\n",
    "\n",
    "    # want a more natural, table-like display\n",
    "    ax.invert_yaxis()\n",
    "    ax.xaxis.tick_top()\n",
    "\n",
    "    # so labels of x and y do not show up and my labels show up instead\n",
    "    ax.set_xticklabels(\n",
    "        list(self.sequence)[segment[0] - self.start_position:segment[1] -\n",
    "                            self.start_position + 1],\n",
    "        fontsize=6.5,\n",
    "        fontname=\"Arial\",\n",
    "        color='k',\n",
    "        minor=False\n",
    "    )\n",
    "    ax.set_yticklabels(\n",
    "        temp_kwargs['neworder_aminoacids'],\n",
    "        fontsize=6,\n",
    "        fontname=\"Arial\",\n",
    "        color=ylabel_color,\n",
    "        minor=False\n",
    "    )\n",
    "\n",
    "    ax2_label = (segment[1] - segment[0] + 1) * ['']\n",
    "    ax2_label[0] = segment[0]\n",
    "    ax2_label[-1] = segment[1]\n",
    "    ax2.set_xticklabels(\n",
    "        ax2_label, fontsize=7, fontname=\"Arial\", color='k', minor=False\n",
    "    )\n",
    "\n",
    "    # align the labels of the y axis\n",
    "    for ylabel in ax.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "\n",
    "    # remove ticks\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "    ax2.yaxis.set_ticks_position('none')\n",
    "    ax2.xaxis.set_ticks_position('none')\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    # return matplotlib object\n",
    "    if temp_kwargs['return_plot_object']:\n",
    "        return fig, ax, ax2\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
